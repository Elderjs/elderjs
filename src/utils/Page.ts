import getUniqueId from './getUniqueId';
import perf from './perf';
import { prepareProcessStack } from './prepareProcessStack';
import { QueryOptions, SettingOptions, ConfigOptions, RequestOptions } from './types';
import { RoutesOptions } from '../routes/types';
import { createReadOnlyProxy } from './createReadOnlyProxy';

const buildPage = async (page) => {
  try {
    page.perf.end('initToBuildGap');
    // stacks
    page.headStack = [];
    page.cssStack = [];
    page.beforeHydrateStack = [];
    page.hydrateStack = [];
    page.customJsStack = [];
    page.footerStack = [];

    await page.runHook('request', page);

    page.perf.start('data');
    if (typeof page.route.data === 'object') {
      page.data = { ...page.data, ...page.route.data };
    } else if (typeof page.route.data === 'function') {
      const dataResponse = await page.route.data({
        data: page.data,
        query: page.query,
        helpers: page.helpers,
        settings: createReadOnlyProxy(page.settings, 'settings', 'data.js'),
        request: createReadOnlyProxy(page.request, 'request', 'data.js'),
        errors: page.errors,
        perf: page.perf,
        allRequests: createReadOnlyProxy(page.allRequests, 'allRequests', 'data.js'),
      });
      if (dataResponse) {
        page.data = dataResponse;
      }
    }
    page.perf.end('data');
    await page.runHook('data', page);

    // start building templates
    page.perf.start('html.template');

    // template building starts here

    const templateHTML = page.route.templateComponent({
      page,
      props: {
        data: page.data,
        helpers: page.helpers,
        settings: page.settings,
        request: page.request,
      },
    });

    page.perf.end('html.template');

    page.perf.start('html.layout');
    const layoutHtml = page.route.layout({
      page,
      props: {
        data: page.data,
        helpers: page.helpers,
        settings: page.settings,
        page,
        templateHTML,
      },
    });
    page.perf.end('html.layout');

    // Run header hooks / stacks to make headString
    await page.runHook('stacks', page);
    page.head = page.processStack('headStack');
    page.cssString = '';
    page.cssString = page.processStack('cssStack');
    page.styleTag = `<style data-name="cssStack">${page.cssString}</style>`;
    page.headString = `${page.head}${page.styleTag}`;
    page.beforeHydrate = page.processStack('beforeHydrateStack');
    page.hydrate = `<script data-name="hydrateStack">${page.processStack('hydrateStack')}</script>`;
    page.customJs = page.processStack('customJsStack');
    page.footer = page.processStack('footerStack');

    await page.runHook('head', page);

    page.perf.start('html.createHtmlString');
    page.htmlString = `<!DOCTYPE html>
      <html lang="en">
        <head>
          <meta charset="UTF-8" />
          <meta name="viewport" content="width=device-width, initial-scale=1" />
          ${page.headString}
        </head>
        <body class="${page.request.route}">
          ${layoutHtml}
          ${page.hydrateStack.length > 0 ? page.beforeHydrate : ''}
          ${page.hydrateStack.length > 0 ? page.hydrate : ''}
          ${page.customJsStack.length > 0 ? page.customJs : ''}
          ${page.footerStack.length > 0 ? page.footer : ''}
        </body>
      </html>
    `;
    page.perf.end('html.createHtmlString');

    await page.runHook('html', page);

    // disconnect timings so we don't get duplicates on next use of page.
    page.perf.end('page');
    page.perf.stop();

    page.timings = page.perf.timings;

    await page.runHook('requestComplete', page);

    if (page.errors.length > 0) {
      await page.runHook('error', page);
    }

    return page;
  } catch (err) {
    console.log(err, page.permalink);
    page.errors.push(err);
    await page.runHook('error', page);
  }
};

class Page {
  uid: string;
  runHook: (string, Object) => Promise<any>;
  allRequests: Array<RequestOptions>;
  request: RequestOptions;
  settings: ConfigOptions & SettingOptions;
  helpers: {};
  data: Object;
  route: any;
  query: QueryOptions;
  errors: any[];
  routes: RoutesOptions;
  processStack: any;
  perf: any;
  customProps: any;
  htmlString: string;
  constructor({
    request,
    settings,
    query,
    helpers,
    data,
    route,
    runHook,
    allRequests,
    routes,
    errors,
    customProps = {},
  }) {
    this.uid = getUniqueId();
    perf(this);
    this.perf.start('page');
    this.perf.start('constructor');
    this.runHook = runHook;
    this.allRequests = allRequests;
    this.request = request;
    this.settings = settings;
    this.helpers = helpers;
    this.data = data;
    this.route = route;
    this.query = query;
    this.errors = [...errors];
    this.routes = routes;
    this.customProps = customProps;
    this.htmlString = '';

    this.processStack = prepareProcessStack(this);

    this.runHook('modifyCustomProps', this);

    this.perf.end('constructor');

    this.perf.start('initToBuildGap');
  }

  build() {
    return buildPage(this);
  }

  html() {
    if (this.htmlString) {
      return this.htmlString;
    }
    return buildPage(this)
      .then((page) => page.htmlString)
      .catch((e) => JSON.stringify(e));
  }
}

export default Page;
