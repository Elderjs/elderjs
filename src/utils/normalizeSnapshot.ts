import windowsPathFix from './windowsPathFix';

const normalizeSnapshot = (val) => {
  if (Object.prototype.toString.call(val) === '[object String]') {
    return windowsPathFix(val);
  }
  if (Object.prototype.toString.call(val) === '[object Array]') {
    return val.map(normalizeSnapshot);
  }
  if (Object.prototype.toString.call(val) === '[object Object]') {
    return Object.keys(val).reduce((out, cv) => {
      // eslint-disable-next-line no-param-reassign
      out[cv] = normalizeSnapshot(val[cv]);
      return out;
    }, {});
  }
  return val;
};

export default normalizeSnapshot;
