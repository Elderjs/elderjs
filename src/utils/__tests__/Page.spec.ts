import Page from '../Page.js';
import normalizeSnapshot from '../normalizeSnapshot.js';
import { describe, it, expect, vi, beforeAll, beforeEach } from 'vitest';
beforeAll(() => {
  vi.resetModules();
});

beforeEach(() => {
  vi.resetModules();
});
vi.mock('../getUniqueId', () => ({ default: () => 'xxxxxxxxxx' }));
vi.mock('../prepareProcessStack', () => ({
  default: (page) => (stackName) => {
    const data = {
      headStack: 'headStack',
      cssStack: 'cssStack',
      hydrateStack: 'hydrateStack',
      beforeHydrateStack: 'beforeHydrateStack',
      customJsStack: 'customJsStack',
      footerStack: 'footerStack',
    };
    if (!page.hydrateStack || !page.hydrateStack.length) {
      // should be done in stacks hook
      page.footerStack = ['footerStack'];
      page.customJsStack = ['customJsStack'];
      page.hydrateStack = ['hydrateStack'];
    }
    if (data[stackName]) {
      return data[stackName];
    }
    return '';
  },
}));
vi.mock('../perf', () => ({
  default: (page) => {
    page.perf = {
      timings: [],
      start: vi.fn(),
      end: vi.fn(),
      stop: vi.fn(),
      prefix: vi.fn(),
    };
  },
}));

const allRequests = [
  {
    slug: 'ash-flat',
    random: 94,
    state: {
      id: 6,
      slug: 'arkansas',
    },
    route: 'cityNursingHomes',
    type: 'build',
    permalink: '/arkansas/ash-flat-nursing-homes/',
  },
  {
    slug: 'albertville',
    random: 33,
    state: {
      id: 4,
      slug: 'alabama',
    },
    route: 'cityNursingHomes',
    type: 'build',
    permalink: '/alabama/albertville-nursing-homes/',
  },
];

const request = {
  slug: 'ash-flat',
  random: 94,
  state: {
    id: 6,
    slug: 'arkansas',
  },
  route: 'cityNursingHomes',
  type: 'build',
  permalink: '/arkansas/ash-flat-nursing-homes/',
};

const settings = {
  distDir: 'test',
  server: false,
  build: {
    shuffleRequests: false,
    numberOfWorkers: -1,
  },
  locations: {
    public: './public/',
    svelte: {
      ssrComponents: './___ELDER___/compiled/',
      clientComponents: './public/dist/svelte/',
    },
    intersectionObserverPoly: '/dist/static/intersection-observer.js',
    buildFolder: '',
    srcFolder: './src/',
    assets: './public/dist/static/',
  },
  debug: {
    stacks: false,
    hooks: true,
    performance: false,
    build: false,
    automagic: false,
  },
  props: {},
  hooks: {
    disable: [],
  },
  plugins: {
    'elder-plugin-upload-s3': {
      dataBucket: 'elderguide.com',
      htmlBucket: 'elderguide.com',
      deployId: '11111111',
    },
  },
  typescript: false,
  $$internal: {
    distElder: 'test/_elderjs',
    hashedComponents: {
      AutoComplete: 'entryAutoComplete',
      Footer: 'entryFooter',
      Header: 'entryHeader',
      HeaderAutoComplete: 'entryHeaderAutoComplete',
      Home: 'entryHomeAutoComplete',
      HomeAutoComplete: 'entryHomeAutoComplete',
      Modal: 'entryModal',
    },
  },
  worker: true,
  cacheBustingId: 'XmjyWGYZtV',
};

const query = {
  db: {
    db: {},
    pool: {
      _events: {},
      _eventsCount: 0,
      options: {
        connectionString: 'postgresql://user:user@localhost:5099/db',
        max: 10,
        idleTimeoutMillis: 10000,
      },
      _clients: [],
      _idle: [],
      _pendingQueue: [],
      ending: false,
      ended: false,
    },
    cnString: {
      connectionString: 'postgresql://user:user@localhost:5099/db',
    },
  },
};

const helpers = {
  permalinks: {},
  metersInAMile: 0.00062137119224,
};

const route = {
  hooks: [],
  template: 'Content.svelte',
  data: () => Promise.resolve({ worldPopulation: 7805564950 }),
  templateComponent: vi.fn(),
  layout: 'Layout.svelte',
  layoutComponent: vi.fn(() => '<div class="container"></div>'),
  parent: 'home',
  $$meta: {
    type: 'route',
    addedBy: 'routejs',
  },
};

const routes = {
  content: {
    hooks: [],
    template: 'Content.svelte',
    $$meta: {
      type: 'route',
      addedBy: 'routejs',
    },
  },
  home: {
    hooks: [
      {
        hook: 'data',
        name: 'testToData',
        description: 'Adds test to data object',
        priority: 50,
      },
    ],
    template: 'Home.svelte',
    $$meta: {
      type: 'route',
      addedBy: 'routejs',
    },
  },
};

describe('#Page', () => {
  const hooks = [];
  const runHook = (hookName) => {
    hooks.push(hookName);
  };

  const pageInput = {
    allRequests,
    request,
    settings,
    query,
    helpers,
    data: {},
    route,
    routes,
    errors: [],
    runHook,
    shortcodes: [],
    perf: {
      start: () => {},
      stop: () => {},
    },
  };

  it('initialize and build', async () => {
    const page = new Page(pageInput);
    expect(normalizeSnapshot(page)).toMatchSnapshot();
    await page.build();

    expect(hooks).toEqual([
      'request',
      'data',
      'shortcodes',
      'stacks',
      'head',
      'compileHtml',
      'html',
      'requestComplete',
    ]);
    expect(normalizeSnapshot(page)).toMatchSnapshot();
  });
});
