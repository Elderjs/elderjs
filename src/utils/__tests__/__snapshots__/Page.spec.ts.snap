// Vitest Snapshot v1

exports[`#Page > initialize and build 1`] = `
{
  "allRequests": [
    {
      "permalink": "/arkansas/ash-flat-nursing-homes/",
      "random": 94,
      "route": "cityNursingHomes",
      "slug": "ash-flat",
      "state": {
        "id": 6,
        "slug": "arkansas",
      },
      "type": "build",
    },
    {
      "permalink": "/alabama/albertville-nursing-homes/",
      "random": 33,
      "route": "cityNursingHomes",
      "slug": "albertville",
      "state": {
        "id": 4,
        "slug": "alabama",
      },
      "type": "build",
    },
  ],
  "beforeHydrateStack": [],
  "bodyAttributesStack": [],
  "bodyAttributesString": "",
  "componentsToHydrate": [],
  "cssStack": [],
  "cssString": "",
  "customJsStack": [],
  "data": {},
  "errors": [],
  "footerStack": [],
  "footerString": "",
  "head": "",
  "headStack": [],
  "headString": "",
  "helpers": {
    "metersInAMile": 0.00062137119224,
    "permalinks": {},
  },
  "htmlAttributesStack": [],
  "htmlAttributesString": "",
  "htmlString": "",
  "hydrateStack": [],
  "moduleJsStack": [],
  "moduleStack": [],
  "next": [Function],
  "perf": {
    "end": [Function],
    "prefix": [Function],
    "start": [Function],
    "stop": [Function],
    "timings": [],
  },
  "processStack": [Function],
  "query": {
    "db": {
      "cnString": {
        "connectionString": "postgresql://user:user@localhost:5099/db",
      },
      "db": {},
      "pool": {
        "_clients": [],
        "_events": {},
        "_eventsCount": 0,
        "_idle": [],
        "_pendingQueue": [],
        "ended": false,
        "ending": false,
        "options": {
          "connectionString": "postgresql://user:user@localhost:5099/db",
          "idleTimeoutMillis": 10000,
          "max": 10,
        },
      },
    },
  },
  "request": {
    "permalink": "/arkansas/ash-flat-nursing-homes/",
    "random": 94,
    "route": "cityNursingHomes",
    "slug": "ash-flat",
    "state": {
      "id": 6,
      "slug": "arkansas",
    },
    "type": "build",
  },
  "resNext": [Function],
  "route": {
    "$$meta": {
      "addedBy": "routejs",
      "type": "route",
    },
    "data": [Function],
    "hooks": [],
    "layout": "Layout.svelte",
    "layoutComponent": [Function],
    "parent": "home",
    "template": "Content.svelte",
    "templateComponent": [Function],
  },
  "routes": {
    "content": {
      "$$meta": {
        "addedBy": "routejs",
        "type": "route",
      },
      "hooks": [],
      "template": "Content.svelte",
    },
    "home": {
      "$$meta": {
        "addedBy": "routejs",
        "type": "route",
      },
      "hooks": [
        {
          "description": "Adds test to data object",
          "hook": "data",
          "name": "testToData",
          "priority": 50,
        },
      ],
      "template": "Home.svelte",
    },
  },
  "runHook": [Function],
  "settings": {
    "$$internal": {
      "distElder": "test/_elderjs",
      "hashedComponents": {
        "AutoComplete": "entryAutoComplete",
        "Footer": "entryFooter",
        "Header": "entryHeader",
        "HeaderAutoComplete": "entryHeaderAutoComplete",
        "Home": "entryHomeAutoComplete",
        "HomeAutoComplete": "entryHomeAutoComplete",
        "Modal": "entryModal",
      },
    },
    "build": {
      "numberOfWorkers": -1,
      "shuffleRequests": false,
    },
    "cacheBustingId": "XmjyWGYZtV",
    "debug": {
      "build": false,
      "hooks": true,
      "performance": false,
      "stacks": false,
    },
    "distDir": "test",
    "hooks": {
      "disable": [],
    },
    "locations": {
      "assets": "./public/dist/static/",
      "buildFolder": "",
      "intersectionObserverPoly": "/dist/static/intersection-observer.js",
      "public": "./public/",
      "srcFolder": "./src/",
      "svelte": {
        "clientComponents": "./public/dist/svelte/",
        "ssrComponents": "./___ELDER___/compiled/",
      },
    },
    "plugins": {
      "elder-plugin-upload-s3": {
        "dataBucket": "elderguide.com",
        "deployId": "11111111",
        "htmlBucket": "elderguide.com",
      },
    },
    "props": {},
    "server": false,
    "typescript": false,
    "worker": true,
  },
  "shortcodes": [],
  "shouldSkipRequest": false,
  "styleTag": "",
  "svelteCss": [],
  "timings": [],
  "uid": "xxxxxxxxxx",
}
`;

exports[`#Page > initialize and build 2`] = `
{
  "allRequests": [
    {
      "permalink": "/arkansas/ash-flat-nursing-homes/",
      "random": 94,
      "route": "cityNursingHomes",
      "slug": "ash-flat",
      "state": {
        "id": 6,
        "slug": "arkansas",
      },
      "type": "build",
    },
    {
      "permalink": "/alabama/albertville-nursing-homes/",
      "random": 33,
      "route": "cityNursingHomes",
      "slug": "albertville",
      "state": {
        "id": 4,
        "slug": "alabama",
      },
      "type": "build",
    },
  ],
  "beforeHydrateStack": [],
  "bodyAttributesStack": [],
  "bodyAttributesString": "",
  "componentsToHydrate": [],
  "cssStack": [],
  "cssString": "cssStack",
  "customJsStack": [
    "customJsStack",
  ],
  "data": {
    "worldPopulation": 7805564950,
  },
  "errors": [],
  "footerStack": [
    "footerStack",
  ],
  "footerString": "
    beforeHydrateStack
    hydrateStack
    customJsStack
    footerStack
    ",
  "head": "headStack",
  "headStack": [],
  "headString": "headStack<style>cssStack</style>",
  "helpers": {
    "metersInAMile": 0.00062137119224,
    "permalinks": {},
  },
  "htmlAttributesStack": [],
  "htmlAttributesString": "",
  "htmlString": "",
  "hydrateStack": [
    "hydrateStack",
  ],
  "layoutHtml": "<div class=\\"container\\"></div>",
  "moduleJsStack": [],
  "moduleStack": [],
  "next": [Function],
  "perf": {
    "end": [Function],
    "prefix": [Function],
    "start": [Function],
    "stop": [Function],
    "timings": [],
  },
  "processStack": [Function],
  "query": {
    "db": {
      "cnString": {
        "connectionString": "postgresql://user:user@localhost:5099/db",
      },
      "db": {},
      "pool": {
        "_clients": [],
        "_events": {},
        "_eventsCount": 0,
        "_idle": [],
        "_pendingQueue": [],
        "ended": false,
        "ending": false,
        "options": {
          "connectionString": "postgresql://user:user@localhost:5099/db",
          "idleTimeoutMillis": 10000,
          "max": 10,
        },
      },
    },
  },
  "request": {
    "permalink": "/arkansas/ash-flat-nursing-homes/",
    "random": 94,
    "route": "cityNursingHomes",
    "slug": "ash-flat",
    "state": {
      "id": 6,
      "slug": "arkansas",
    },
    "type": "build",
  },
  "resNext": [Function],
  "route": {
    "$$meta": {
      "addedBy": "routejs",
      "type": "route",
    },
    "data": [Function],
    "hooks": [],
    "layout": "Layout.svelte",
    "layoutComponent": [Function],
    "parent": "home",
    "template": "Content.svelte",
    "templateComponent": [Function],
  },
  "routes": {
    "content": {
      "$$meta": {
        "addedBy": "routejs",
        "type": "route",
      },
      "hooks": [],
      "template": "Content.svelte",
    },
    "home": {
      "$$meta": {
        "addedBy": "routejs",
        "type": "route",
      },
      "hooks": [
        {
          "description": "Adds test to data object",
          "hook": "data",
          "name": "testToData",
          "priority": 50,
        },
      ],
      "template": "Home.svelte",
    },
  },
  "runHook": [Function],
  "settings": {
    "$$internal": {
      "distElder": "test/_elderjs",
      "hashedComponents": {
        "AutoComplete": "entryAutoComplete",
        "Footer": "entryFooter",
        "Header": "entryHeader",
        "HeaderAutoComplete": "entryHeaderAutoComplete",
        "Home": "entryHomeAutoComplete",
        "HomeAutoComplete": "entryHomeAutoComplete",
        "Modal": "entryModal",
      },
    },
    "build": {
      "numberOfWorkers": -1,
      "shuffleRequests": false,
    },
    "cacheBustingId": "XmjyWGYZtV",
    "debug": {
      "build": false,
      "hooks": true,
      "performance": false,
      "stacks": false,
    },
    "distDir": "test",
    "hooks": {
      "disable": [],
    },
    "locations": {
      "assets": "./public/dist/static/",
      "buildFolder": "",
      "intersectionObserverPoly": "/dist/static/intersection-observer.js",
      "public": "./public/",
      "srcFolder": "./src/",
      "svelte": {
        "clientComponents": "./public/dist/svelte/",
        "ssrComponents": "./___ELDER___/compiled/",
      },
    },
    "plugins": {
      "elder-plugin-upload-s3": {
        "dataBucket": "elderguide.com",
        "deployId": "11111111",
        "htmlBucket": "elderguide.com",
      },
    },
    "props": {},
    "server": false,
    "typescript": false,
    "worker": true,
  },
  "shortcodes": [],
  "shouldSkipRequest": false,
  "styleTag": "<style>cssStack</style>",
  "svelteCss": [],
  "templateHtml": undefined,
  "timings": [],
  "uid": "xxxxxxxxxx",
}
`;
